text/0000-cargo-asymmetric-tokens.md

```diff
@@ -0,0 +1,183 @@
+
+> If a user misconfigures a token to go to the wrong registry (typosquatting, homoglyph, or copy-paste error), then the recipient has the token.
+
+The token includes the URL so the signature is only valid for that URL, the token is not valid for the real registry.
+
+> If a registry does not adequately protect its copy of the tokens then a database disclosure can leak all the users' tokens. ([cc: crates.io security advisory](https://blog.rust-lang.org/2020/07/14/crates-io-security-advisory.html))
+
+There is no reason for the registry to even see the private key. Even if the registry wants to generate keys for its users there is no need to store private keys. Disclosure of public keys is not a security risk, as they can not be used to sign new tokens.
+
+> Fundamentally these are all problems only because once an attacker has seen a secret token they have all that is needed to act on that user's behalf.
+
+Without the private key an asymmetric cryptography token can only be used for the intended registry, for the intended action, and for a limited amount of time. This mitigates the risk of disclosure.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+This gets Cargo involved in the cryptographic standards used by registries, which puts a lot of complexity on ourselves. Now rust teams need to be involved in conversations about what cryptographic standards alternative registries choose to use.
+
+Furthermore this RFC attempts to make a start on solving several problems at the same time. It may be that in time we discover these problems need to be solved separately. If we end up with a separate system for code signing and a separate system for authorization, then a simpler more direct method of authentication might have been a better choice.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+Continue with the existing secret tokens. Private registries that want to provide this kind of functionality can create a bespoke system for their exact needs. For example, only generating short-lived tokens and having the user log in daily.
+In practice, I suspect many registries will not, leading to an ecosystem where most registries use less secure authentication, and creating more hazards for users. Some security properties (e.g. not supporting tokens from one registry on another) work better when all registries support them.
+
+We could use PASETO `v4.public` instead of `v3.public`. The `v4` standard uses more modern crypto, based on Ed25519.
+Unfortunately, most existing hardware tokens do not support Ed25519. By using `v3.public` based on P-384 we allow a `credential-process` to keep keys on the hardware.
```

Comment:
I found some comparison on hn, which says that ed25519 is generally better: https://news.ycombinator.com/item?id=19749452

Furthermore, ed25519 has less red stuff at the table at [safecurves](http://safecurves.cr.yp.to/index.html) compared to the NIST curves.

Generally it might be a good idea to support multiple schemes from the start, to ensure that there is algorithm agility.

Reply:
> Furthermore, ed25519 has less red stuff at the table at [safecurves](http://safecurves.cr.yp.to/index.html) compared to the NIST curves.

The table on the safecurves site is critiquing the obsolete and incomplete formulas I was describing earlier, and is specifically addressed in the paper I linked (p.5), where citation `[14]` is the safecurves site:

> Prime order curves can be safe. Several of the standardized prime order curves mentioned
above have recently been critiqued in [14], where they were deemed not to meet (some or all
of) the four “ECC security” requirements: (i) Ladder, (ii) Twists, (iii) Completeness, and (iv)
Indistinguishability.
> On the contrary, this paper shows that prime order curves have complete formulas that are
comparably efficient.

Why the safecurves site has not been updated to reflect the developments of this paper is uncertain, however it contains obsolete information.

The security properties of implementations of Edwards curves as used by Ed25519 or short Weierstrass curves like NIST P-256 and P-384 implemented using the complete projective formulas are effectively the same, and algorithmically quite similar. As it were, the RustCrypto [`p256`] (https://github.com/RustCrypto/elliptic-curves/tree/master/p256) crate is implemented using these formulas, and we are working implementing [`p384`](https://github.com/RustCrypto/elliptic-curves/tree/master/p384) with them as well.

Where Ed25519 benefits is areas like performance (Edwards curves are more amenable to parallel implementations) and more exotic applications like batch verification, multisignatures, and signature aggregation, which are outside of the scope of this particular use case.

> Generally it might be a good idea to support multiple schemes from the start, to ensure that there is algorithm agility.

I would suggest supporting Ed25519 along with ECDSA/P-{256,384} (possibly only choosing one of P-256 or P-384). While Ed25519 is great, support for it in things like hardware tokens, OS keychains, HSMs, and cloud KMS services is still somewhat spotty, but slowly improving.

Reply:
To make sure I understand correctly, the RFC as-is currently would *not* support Ed25519, and supporting both v3 and v4 (and thus both P-384 and Ed25519) is not part of the RFC? But the RFC is written in a way that that would be a possible extension in the future?

Reply:
I am not a cryptographer and am new to all the details here. So I'm going to explain as carefully as I can, so that someone can tell me if I have something wrong.

`PASETO V4` is an encoding that requires the use of the algorithm `Ed25519`.
`PASETO V3` is an encoding that requires the use of the algorithm `ECDSA/P-384`.
Also all the PASETO variations clearly identify what version they came from.

The RFC as currently written only supports `PASETO V3`. That means that cargo will only understand how to generate and read keys for `PASETO V3` and use the algorithm `ECDSA/P-384`. Furthermore it will insist that credential processes return something that appears to be correctly formatted as a `PASETO V3`.

It is not hard to imagine how this could be lifted in a future RFC. This would allow us to allow for a new format, say `PASETO V4` or `Biscuit`.
That RFC would have to specify:
- How `--generate-keypair` decides `V3`/`V4`/`Biscuit`? Presumably, a new flag to specify format and a timeline for the default to change.
- When reading a config file how cargo decides which one it has found? Almost certainly, read the version number encoded in the key and use its matching format.
- When receiving from a credential process which style to check? Presumably, we allow anything that looks like any of the supported formats. (Note that this could be done before and independently of the other changes.)
- Whether to deprecate and remove the old format? This would require extreme justification, but maybe worth discussion depending on the urgency of switching.
- Whether and how a registry can request a particular format? Presumably, Cargo will send whatever format is specified in the config, if it is not a format supported by this registry the registry will return an error code to show the problem. (How does this work with credential providers?)

Does that answer your questions? Did I miss something important?

Reply:
Yes, that all makes sense, thanks. I am not clear on how 'format detection' would be done (are all these formats clearly distinct from each other in their encoded form, do we use filenames, whatever), but as long as that option exists I do not care about the technical details. :)

Reply:
> are all these formats clearly distinct from each other in their encoded form?

Basically yes. Tokens encoded in `PASETO V3` start with `v3.public.`. `PASETO V3` keys encoded in `PASERK` start with `k3.secret.` and key IDs start with `k3.pid.`. (That sentence is also true if you replace all the `3`s with `4`s.) If we are switching to a different family of formats, which do not have a clear prefix to identify them, we can define that we are sticking some arbitrary string in front for disambiguation purposes.